examples/typing/mg.jl

```diff
@@ -3,6 +3,149 @@ using Cassette
 using LightGraphs
 using MetaGraphs
 
+Cassette.@context TypeCtx
+            
+"""   TypeCtx
+
+creates a MetaDiGraph tracking the types of args and ret values throughout a script
+
+"""
+TypeCtx
+            
+            
+
+"""   FCollector(depth::Int,frame::function,data::FCollector)
+
+struct to collect all the "frames" called throughout a script
+        
+"""
+mutable struct FCollector{I,F,C}
+    depth::I
+    frame::F
+    data::Vector{C}
+end
+            
+
+
+"""   FCollector(depth::Int,frame::Frame)
+
+this is an initialization funtion for the FCollector
+
+"""
+function FCollector(d::Int, f)
+    FCollector(d, f, FCollector[])
+end
+
+""" Frame(func, args, ret, subtrace)
+
+a structure to hold metadata for recursive type information for each function call
+Every frame can be thought of as a single stack frame when a function is called
+            
+"""
+mutable struct Frame{F,T,U}
+    func::F
+    args::T
+    ret::U
+end
+            
+function Cassette.overdub(ctx::TypeCtx, f, args...) # add boilerplate for functionality
+    c = FCollector(ctx.metadata.depth-1, Frame(f, typeof.(args), Any))
+    push!(ctx.metadata.data, c)
+    if c.depth > 0 && Cassette.canrecurse(ctx, f, args...)
+        newctx = Cassette.similarcontext(ctx, metadata = c)
+        z = Cassette.recurse(newctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    else
+        z = Cassette.fallback(ctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    end
+end
+
+Cassette.canrecurse(ctx::Type{TypeCtx},::typeof(Base.vect), args...) = false # limit the stacktrace in terms of which to recurse on
+Cassette.canrecurse(ctx::Type{TypeCtx},::typeof(FCollector)) = false
+function Cassette.canrecurse(ctx::TypeCtx,
```

Comment:
adding to this list more of the lower level functions would reduce the verbosity of the KGs you get